src/DFAMin.hs

```diff
@@ -112,57 +131,47 @@ groupEquivStates DFA { dfa_states = statemap }
     accept_groups :: [EquivalenceClass]
     accept_groups = map IS.fromList (Map.elems accept_map)
 
-    init_p, init_q :: [EquivalenceClass]
-    init_p  -- Issue #71: each EquivalenceClass needs to be a non-empty set
-      | IS.null nonaccepting_states = accept_groups
-      | otherwise                   = nonaccepting_states : accept_groups
+    init_r, init_q :: [EquivalenceClass]
+    init_r  -- Issue #71: each EquivalenceClass needs to be a non-empty set
+      | IS.null nonaccepting_states = []
+      | otherwise                   = [nonaccepting_states]
     init_q = accept_groups
 
-    -- map token T to
-    --   a map from state S to the list of states that transition to
+    -- a map from token T to
+    --   a map from state S to the set of states that transition to
     --   S on token T
-    -- This is a cache of the information needed to compute x below
-    bigmap :: IntMap (IntMap [SNum])
-    bigmap = IM.fromListWith (IM.unionWith (++))
-                [ (i, IM.singleton to [from])
+    -- This is a cache of the information needed to compute xs below
+    bigmap :: IntMap (IntMap EquivalenceClass)
+    bigmap = IM.fromListWith (IM.unionWith IS.union)
+                [ (i, IM.singleton to (IS.singleton from))
                 | (from, state) <- Map.toList statemap,
                   (i,to) <- IM.toList (state_out state) ]
 
-    -- incoming I A = the set of states that transition to a state in
-    -- A on token I.
-    incoming :: Int -> IntSet -> IntSet
-    incoming i a = IS.fromList (concat ss)
-       where
-         map1 = IM.findWithDefault IM.empty i bigmap
-         ss = [ IM.findWithDefault [] s map1
-              | s <- IS.toList a ]
-
-    -- The outer loop: recurse on each set in Q
-    go ::  [EquivalenceClass] -> [EquivalenceClass] -> [EquivalenceClass]
-    go p [] = p
-    go p (a:q) = go1 0 p q
-     where
-       -- recurse on each token (0..255)
-       go1 256 p q = go p q
-       go1 i   p q = go1 (i+1) p' q'
+    -- The outer loop: recurse on each set in R and Q
+    go :: [EquivalenceClass] -> [EquivalenceClass] -> [EquivalenceClass]
+    go r [] = r
+    go r (a:q) = uncurry go $ List.foldl' go0 (a:r,q) xs
+      where
+        xs :: [EquivalenceClass]
+        xs = filter (not . IS.null)
+           . map (\m -> IS.unions [IM.findWithDefault IS.empty s m | s <- IS.toList a])
+           $ IM.elems bigmap
```

Comment:
OK first rewrite, but I am not done yet, just trying to take small steps
```suggestion
        xs = 
          [ x
          | preimageMap <- IM.elems bigmap -- map per token
          , let x = IS.unions [IM.findWithDefault IS.empty s preimageMap | s <- IS.toList a])
          , not $ IS.null x
          ]
```

Reply:
I think I prefer using `restrictKeys` for clarity, but please make the `restrictKeys` CPP use `MIN_VERSION_containers`.
```suggestion
        xs = 
          [ x
          | preimageMap <- IM.elems bigmap -- map per token
          , let x = IS.unions $ IM.elems $ restrictKeys preimageMap a
          , not $ IS.null x
          ]
```

Reply:
I should ask @treeowl, but I can imagine making a `Map -> Set -> List` function expressly for the fusion of  `IM.toList (restrictKeys _ _)` and similar, so that this code really does become the optimal version.

Reply:
Yes, it makes sense to fuse restriction with list conversion. Is that worth rewrite rules for `containers`? I'm not sure. What do you think? You can certainly pull in the internals and roll your own.

Reply:
To save you a little time, Haddock tells me it's`MIN_VERSION_containers(0, 5, 8)`

Reply:
Another option might be yet another variant of `mergeA` combining sets with maps. Then you could work in `Const (Endo [a])` or whatever.

Reply:
Thanks @treeowl. That last part sounds much nicer than the combinatorial explosion of manual fusion functions one would need.

Reply:
Let me describe why I discouraged to use `restrictKeys`
> Time complexity of restrictKeys is O(n+m) while time complexity of repeated use of findWithDefault is O(min(n,W)*m).
In Hopcroft's algorithm, n is the number of all states and m is the number of states in each equivalence class.

`n` is number of all states that is pretty big.
`m` is the number of states in each equivalence class.
That means, in the end of the algorithm, `m` might be 1 or small number if the input DFA was already almost optimized. Then the effectual time complexity would be `O(n)` for `restrictKeys` and `O(W*m)` for `findWithDefault`. So, `restrictKeys` might be slower then the repeated `findWithDefault`.

Reply:
I don't think that's right? Haddock says `restrictKeys` is `O(m*log(n/m + 1)), m <= n`, and `findWithDefault` is `O(log n)`. So I that means `O(log n)` vs `O(n log n)` when `n >> m`. Intuitively this makes sense. With the repeated lookups, we have to traverse through the same intermediate nodes again and again. But with `restrictKeys`, we do everything in one pass.

Reply:
BTW, we don't really need `IM.elems` because `IS.unions` works with `Foldable` since containers-0.6.0.
List structure also have to be fused to be optimal.
```suggestion
        xs = filter (not . IS.null)
           . map (IS.unions . flip IM.restrictKeys a)
           $ IM.elems bigmap
```

Reply:
@Ericson2314 You have to see `IntMap.restrictKeys`, not `Map.restrictKeys`. They have different big-O